Kernel/Arch/riscv64/Interrupts.cpp

```diff
@@ -32,6 +35,90 @@ void dump_registers(RegisterState const& regs)
     dbgln("s1={:p} s2={:p} s3={:p} s4={:p} s5={:p} s6={:p} s7={:p} s8={:p} s9={:p} s10={:p} s11={:p}", regs.x[8], regs.x[17], regs.x[18], regs.x[19], regs.x[20], regs.x[21], regs.x[22], regs.x[23], regs.x[24], regs.x[25], regs.x[26]);
 }
 
+extern "C" void trap_handler(TrapFrame& trap_frame);
+extern "C" void trap_handler(TrapFrame& trap_frame)
+{
+    auto scause = trap_frame.regs->scause;
+
+    if ((to_underlying(scause) & RISCV64::CSR::SCAUSE_INTERRUPT_MASK) != 0) {
+        // Interrupt
+
+        Processor::current().enter_trap(trap_frame, true);
+
+        auto interrupt_number = to_underlying(scause) & ~RISCV64::CSR::SCAUSE_INTERRUPT_MASK;
+
+        auto* handler = s_interrupt_handlers[interrupt_number];
+        VERIFY(handler);
+        handler->increment_call_count();
+        handler->handle_interrupt(*trap_frame.regs);
+        handler->eoi();
```

Comment:
Is this EOI thing another x86-ism? I would expect that clearing the interrupt at the interrupt source is the responsibility of the handler function.

Reply:
Not really, interrupt controllers usually want to know when the CPU is done handling an interrupt.
But yeah, we need to rethink how to organize interrupt controllers and interrupt handling more cleanly, as certain interrupt controllers might not need an eoi or need other interfaces.